File: sdk-python/examples/ciba_polling_negative_path.py
from __future__ import annotations

import time

from nexusid_sdk import create_auth_api, create_client
from nexusid_sdk.errors import APIResponseError

BASE_URL = "http://127.0.0.1:4000"
CLIENT_ID = "enterprise-client"
CLIENT_SECRET = "replace-with-client-secret"


def oauth_error_code(error: APIResponseError) -> str | None:
    if isinstance(error.details, dict):
        code = error.details.get("error")
        if isinstance(code, str):
            return code
    return None


def main() -> None:
    auth = create_auth_api(create_client(base_url=BASE_URL))

    ciba_start = auth.start_ciba_authentication(
        client_id=CLIENT_ID,
        client_secret=CLIENT_SECRET,
        login_hint="user@example.com",
        scope=["openid", "profile"],
        binding_message="Approve sign-in from support console",
    )

    auth_req_id = str(ciba_start["auth_req_id"])
    delay_seconds = int(ciba_start.get("interval", 5))

    for attempt in range(1, 9):
        time.sleep(delay_seconds)

        try:
            token = auth.exchange_ciba_token(
                client_id=CLIENT_ID,
                client_secret=CLIENT_SECRET,
                auth_req_id=auth_req_id,
            )
            print("CIBA approved; access token:", token["access_token"])
            return
        except APIResponseError as error:
            oauth_error = oauth_error_code(error)

            if oauth_error == "authorization_pending":
                print(f"Attempt {attempt}: authorization is still pending.")
                continue

            if oauth_error == "slow_down":
                delay_seconds += 1
                print(f"Attempt {attempt}: server requested slower polling. New interval {delay_seconds}s.")
                continue

            # access_denied, expired_token, invalid_grant, or any other terminal error
            raise

    print("No terminal result yet; continue polling or request user action.")


if __name__ == "__main__":
    main()

File: sdk/examples/ciba-polling-negative-path.ts
import {
  APIResponseError,
  createAuthAPI,
  createClient
} from "../src/index.js";

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const getOAuthErrorCode = (error: APIResponseError): string | undefined => {
  const details = error.details as { error?: string } | undefined;
  return details?.error;
};

async function main() {
  const baseUrl = "https://iam.example.com";
  const auth = createAuthAPI(createClient({ baseUrl }));

  const cibaStart = await auth.startCibaAuthentication({
    clientId: "enterprise-client",
    clientSecret: "replace-with-client-secret",
    loginHint: "user@example.com",
    scope: ["openid", "profile"],
    bindingMessage: "Approve sign-in from support console"
  });

  let delayMs = cibaStart.interval * 1000;

  for (let attempt = 1; attempt <= 8; attempt += 1) {
    await sleep(delayMs);

    try {
      const token = await auth.exchangeCibaToken({
        clientId: "enterprise-client",
        clientSecret: "replace-with-client-secret",
        authReqId: cibaStart.auth_req_id
      });

      console.log("CIBA approved; access token:", token.access_token);
      return;
    } catch (error) {
      if (!(error instanceof APIResponseError)) {
        throw error;
      }

      const oauthError = getOAuthErrorCode(error);

      if (oauthError === "authorization_pending") {
        console.log(`Attempt ${attempt}: authorization is still pending.`);
        continue;
      }

      if (oauthError === "slow_down") {
        delayMs += 1000;
        console.log(`Attempt ${attempt}: server requested slower polling. New interval ${delayMs}ms.`);
        continue;
      }

      // access_denied, expired_token, invalid_grant, or any other terminal error
      throw error;
    }
  }

  console.log("No terminal result yet; continue polling or request user action.");
}

void main();
